features/resources/programs/slot_machine.teal

```diff
@@ -0,0 +1,229 @@
+// TODO: award(byte[3])uint64
+// MAYBE TODO:
+// * _create(int,string[])void
+// * disallow updates, optins, etc.
+
+// Input: None
+// Output: 3 random bytes from each slot of slot machine as well as corresponding witnesses
+
+// 0. On creation, provide s_i , (combo_j, award_j) params and set the following app global vars:
+//     * RandomByte:app_id - global-int
+//     * itob(1):s_1       - global-byte
+//     * itob(2):s_2       - global-byte
+//     ...
+//     * itob(n):s_n       - global-byte
+    
+//     * combo_1:award_1   - global-int
+//     * combo_2:award_1   - global-int
+//     ...
+//     * combo_m:award_m   - global-int
+
+// (at runtime, you can figure out what n is with GlobalNumByteSlice)
+
+// 1. create variables "reels" and "witnesses" each containing ""
+// 2. for i = 1 to n:
+//     c_i, w_i = RandomByte.randElement(s_n)
+//     reels += c_i
+//     witnesses += w_i
+// 3. return reels + witnesses
+
+#pragma version 6
+txn ApplicationID
+bz create
+
+txn ApplicationArgs 0
+method "spin(application,application)(byte[3],byte[17],byte[17],byte[17])"
+==
+assert  // for now, only one method allowed
+
+byte "num_reels"
+app_global_get
+// >>> numReels
+
+dup
+assert // gotta spin something
+
+// spin reel2, then reel1 then reel0
+itxn_begin
+spin:
+    int 1
+    -
+    dup
+    // >>> reelNum <-- reelNum
+
+    int appl
+    itxn_field TypeEnum
+
+    txn Applications 2  // RandomByte
+    itxn_field ApplicationID
+
+    method "randElement(string,application)(byte,byte[17])"
+    itxn_field ApplicationArgs
+
+    byte "reel"
+    swap
+    itob
+    // >>> reelNum <-- "reel" <-- itob(reelNum)
+    concat
+    app_global_get
+    // >> reelNum <-- reelContents
+
+    // pushbytes "reel"
+    // callsub debug
+
+    dup
+    len
+    itob
+    extract 6 0
+    swap
+    concat
+
+    // pushbytes "randElement input"
+    // callsub debug
+
+    itxn_field ApplicationArgs
+    // >>> reelNum
+
+    txn Applications 1 // FakeRandom
+    itxn_field Applications
+    
+    dup
+    // >>> reelNum <-- reelNum
+    bz submit
+itxn_next
+b spin
+
+submit:
+    int 0
+    ==
+    assert
+    // <<<EMPTY>>>
+itxn_submit
+
+gitxna 2 Logs 0
+
+// pushbytes "gitxna 2 Logs 0"
+// callsub debug
+
+// >>> return|randByte2|witness2
+callsub extract_element_and_witness
+// >>> randByte2 <-- witness2
+
+dup2
+pushbytes "witness2"
+callsub debug
+pop
+pushbytes "randByte2"
+callsub debug
+pop
+
+
+gitxna 1 Logs 0
+
+// pushbytes "gitxna 1 Logs 0"
+// callsub debug
+
+callsub extract_element_and_witness
+// >>> randByte2 <-- witness2 <-- randByte1 <-- witness1
+
+dup2
+pushbytes "witness1"
+callsub debug
+pop
+pushbytes "randByte1"
+callsub debug
+pop
+
+
+gitxna 0 Logs 0
+
+// pushbytes "gitxna 0 Logs 0"
+// callsub debug
+
+callsub extract_element_and_witness
+// >>> randByte2 <-- witness2 <-- randByte1 <-- witness1 <-- randByte0 <-- witness0
+
+dup2
+pushbytes "witness0"
+callsub debug
+pop
+pushbytes "randByte0"
+callsub debug
+pop
```

Comment:
The debugging helper is really useful! I suppose it should log some message like
```
dbg [witness_i] --->
** witness_i content **
```
to log, and the whole block of code is to keep track of stuffs before ABI return, is that correct?

Reply:
Yes, that's what it does... but you'll quickly run out of space if you put too many messages in. That's why I have a bunch of them commented out.